server/src/main/java/org/elasticsearch/snapshots/SnapshotShutdownProgressTracker.java

```diff
@@ -0,0 +1,271 @@
+ * which see {@link #SNAPSHOT_PROGRESS_DURING_SHUTDOWN_LOG_INTERVAL_SETTING}.
+ */
+public class SnapshotShutdownProgressTracker {
+
+    /** How frequently shard snapshot progress is logged after receiving local node shutdown metadata. */
+    public static final Setting<TimeValue> SNAPSHOT_PROGRESS_DURING_SHUTDOWN_LOG_INTERVAL_SETTING = Setting.positiveTimeSetting(
+        "snapshots.shutdown.progress.interval",
+        TimeValue.timeValueSeconds(5),
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    private static final Logger logger = LogManager.getLogger(SnapshotShutdownProgressTracker.class);
+
+    private final Supplier<String> getLocalNodeId;
+    private final ThreadPool threadPool;
+
+    private volatile TimeValue progressLoggerInterval;
+    private Scheduler.Cancellable scheduledProgressLoggerFuture;
+
+    /**
+     * The time at which the cluster state update began that found a shutdown signal for this node. Negative value means unset (node is not
+     * shutting down).
+     */
+    private volatile long shutdownStartMillis = -1;
+
+    /**
+     * The time at which the cluster state finished setting shard snapshot states to PAUSING, which the shard snapshot operations will
+     * discover asynchronously. Negative value means unset (node is not shutting down)
+     */
+    private volatile long shutdownFinishedSignallingPausingMillis = -1;
+
+    /**
+     * Tracks the number of shard snapshots that have started on the data node but not yet finished.
+     */
+    private final AtomicLong numberOfShardSnapshotsInProgressOnDataNode = new AtomicLong();
+
+    /**
+     * The logic to track shard snapshot status update requests to master can result in duplicate requests (see
+     * {@link ResultDeduplicator}), as well as resending requests if the elected master changes.
+     * Tracking specific requests uniquely by snapshot ID + shard ID de-duplicates requests for tracking.
+     * Also tracks the absolute start time of registration, to report duration on de-registration.
+     */
+    private final Map<String, Long> shardSnapshotRequests = ConcurrentCollections.newConcurrentMap();
+
+    /**
+     * Track how the shard snapshots reach completion during shutdown: did they fail, succeed or pause?
+     */
+    private final AtomicLong doneCount = new AtomicLong();
+    private final AtomicLong failureCount = new AtomicLong();
+    private final AtomicLong abortedCount = new AtomicLong();
+    private final AtomicLong pausedCount = new AtomicLong();
+
+    public SnapshotShutdownProgressTracker(
+        Supplier<String> localNodeIdSupplier,
+        BiConsumer<Setting<TimeValue>, Consumer<TimeValue>> addSettingsUpdateConsumer,
+        Settings settings,
+        ThreadPool threadPool
+    ) {
+        this.getLocalNodeId = localNodeIdSupplier;
+        this.progressLoggerInterval = SNAPSHOT_PROGRESS_DURING_SHUTDOWN_LOG_INTERVAL_SETTING.get(settings);
```

Comment:
Suggest using `org.elasticsearch.common.settings.AbstractScopedSettings#initializeAndWatch` to avoid having to  set the field explicitly here